### DISCUSSION TITLE
How to add Control in my Action?

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#29567] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

There does not appear to be an AddControl method in FEProblemBase. How do I add a Control in my Action?

**Comment by GiudGiud** [comment]
I think the control warehouse lives on the app instead of the problem. I can check when I get to a computer

**Reply by jessecarterMOOSE** [reply, marked as answer]
Yeah I think I figured it out:
```
#include Control.h
...
InputParameters params = _factory.getValidParams("MyControl");
// set params...
std::shared_ptr<Control> control = _factory.create<Control>("MyControl", "control_name", params);
_problem->getControlWarehouse().addObject(control);
```

**Reply by jessecarterMOOSE** [reply]
borrowed that from `AddControlAction` in the Framework src
